vectordatasource/meta/kinds.py

```diff
@@ -0,0 +1,473 @@
+        'warehouse',
+        'wayside_shrine',
+        'works',
+    ],
+    'mz_building_part_kind_detail': [
+        None,
+        'arch',
+        'balcony',
+        'base',
+        'column',
+        'door',
+        'elevator',
+        'entrance',
+        'floor',
+        'hall',
+        'main',
+        'passageway',
+        'pillar',
+        'porch',
+        'ramp',
+        'roof',
+        'room',
+        'steps',
+        'stilobate',
+        'tier',
+        'tower',
+        'verticalpassage',
+        'wall',
+        'window',
+    ],
+}
+
+
+def all_possible_return_values_of(func):
+    return _RETURN_VALUES[func]
+
+
+# need this to pass into the CSVMatcher, which expects a Shapely geometry.
+FakeShape = namedtuple('FakeShape', 'type')
+
+
+def parse_col_values(col, item):
+    if isinstance(col, (str, unicode)):
+        all_kinds = [col]
+
+    elif col is None:
+        all_kinds = [None]
+
+    elif isinstance(col, list):
+        for c in col:
+            assert isinstance(c, (str, unicode))
+
+        all_kinds = col
+
+    elif isinstance(col, dict):
+        if col.keys() == ['col']:
+            all_kinds = parse_filter_for_col(col['col'], item)
+
+        elif col.keys() == ['case']:
+            all_kinds = parse_case_for_kinds(col['case'], item)
+
+        elif col.keys() == ['call']:
+            all_kinds = all_possible_return_values_of(col['call']['func'])
+
+        else:
+            raise ValueError("Don't understand dict column with keys %r in %r"
+                             % (col.keys(), col))
+
+    else:
+        raise ValueError("Don't understand column %s in %r" % (type(col), col))
+
+    return all_kinds
+
+
+def parse_item(layer_name, item, sort_rank, include_kind_detail):
+    kind = item['output'].get('kind')
+    if not kind:
+        return {}
+
+    all_kinds = parse_col_values(kind, item)
+
+    kind_detail = item['output'].get('kind_detail')
+    if kind_detail:
+        kind_details = parse_col_values(kind_detail, item)
+    else:
+        kind_details = [None]
+
+    values = {}
+    for k in all_kinds:
+        for kind_detail in kind_details:
+            key = KindKey(layer_name, k, kind_detail)
+
+            assert isinstance(k, (str, unicode))
+            assert isinstance(kind_detail, (str, unicode, type(None)))
+
+            # we need to fake up some of this data, so the sort ranks might
+            # not be exactly correct...
+            shape = FakeShape(None)
+            props = {'kind': k, 'kind_detail': kind_detail}
+            zoom = 16
+            result = sort_rank(shape, props, zoom)
+            if result is None:
+                sort_rank_val = None
+            else:
+                sort_rank_key, sort_rank_val = result
+                assert sort_rank_key == 'sort_rank'
+                sort_rank_val = int(sort_rank_val)
+
+            info = KindInfo(parse_start_zoom(item['min_zoom']), sort_rank_val)
+
+            values[key] = info
```

Comment:
Should some of this be predicated on the `include_kind_detail` parameter? If not, to me it looks like that isn't used and maybe the function shouldn't take it?

Reply:
Yes, well spotted, thanks!

That's definitely a bug. I think it should have aggregated by `kind_detail`, but obviously I forgot to do that... oops. Hopefully it's fixed in 6ff5b58 - would appreciate it if you could take a look and let me know what you think.

Reply:
👍